provider/provider.go

```diff
@@ -0,0 +1,270 @@
+	Stop() error
+
+	// ReadResource refreshes a resource and returns its current state.
+	ReadResource(ReadResourceRequest) ReadResourceResponse
+
+	// PlanResourceChange takes the current state and proposed state of a
+	// resource, and returns the planned final state.
+	PlanResourceChange(PlanResourceChangeRequest) PlanResourceChangeResponse
+
+	// ApplyResourceChange takes the planned state for a resource, which may
+	// yet contain unknown computed values, and applies the changes returning
+	// the final state.
+	ApplyResourceChange(ApplyResourceChangeRequest) ApplyResourceChangeResponse
+
+	// ImportResourceState requests that the given resource be imported.
+	ImportResourceState(ImportResourceStateRequest) ImportResourceStateResponse
+
+	// ReadDataSource returns the data source's current state.
+	ReadDataSource(ReadDataSourceRequest) ReadDataSourceResponse
+}
+
+type GetSchemaResponse struct {
+	// Provider is the schema for the provider itself.
+	Provider *configschema.Block
+
+	// ResourceTypes map the resource type name to that type's schema.
+	ResourceTypes map[string]*configschema.Block
+
+	// DataSources maps the data source name to that data source's schema.
+	DataSources map[string]*configschema.Block
+
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ValidateProviderConfigRequest struct {
+	// Config is the complete configuration value for the provider.
+	Config cty.Value
+}
+
+type ValidateProviderConfigResponse struct {
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ValidateResourceTypeConfigRequest struct {
+	// TypeName is the name of the resource type to validate.
+	TypeName string
+
+	// Config is the configuration value to validate, which may contain unknown
+	// values.
+	Config cty.Value
+}
+
+type ValidateResourceTypeConfigResponse struct {
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type ValidateDataSourceConfigRequest struct {
+	// TypeName is the name of the data source type to validate.
+	TypeName string
+
+	// Config is the configuration value to validate, which may contain unknown
+	// values.
+	Config cty.Value
+}
+
+type ValidateDataSourceConfigResponse struct {
+	// Diagnostics contains any warnings or errors from the method call.
+	Diagnostics tfdiags.Diagnostics
+}
+
+type UpgradeResourceStateRequest struct {
+	// TypeName is the name of the resource type being upgraded
+	TypeName string
+
+	// Version is version of the schema that created the current state.
+	Version int
+
+	// PriorStateRaw is the state that needs to be upgraded to match the current
+	// schema version. Because the schema is unknown, this contains only the
+	// raw bytes as stored in the state. It is up to the provider to interpret
+	// the payload and return a state encoded with the current schema.
+	PriorStateRaw []byte
```

Comment:
There are two different cases the provider needs to deal with here: a `[]byte` containing JSON and a `map[string]string` containing flatmap encoding.

While we could also send the flatmap encoding over here as JSON, the provider still needs a way to distinguish which case it's dealing with, so maybe simpler to echo how the underlying state structs are built here:

```go
    PriorStateJSON []byte
    PriorStateFlatmap map[string]string
```

...where the caller guarantees that exactly one of these is set and the other is always `nil`.

We can't do the flatmap shimming in Core before calling this method because Core doesn't have the schema for the old version available.

Reply:
I was going to leave that entirely up to the provider, making this just a transport for the raw state, but now that you mention it I think I'll add a `PriorState` message with multiple encoding types right in the protocol to remove any ambiguity.

Reply:
Oh I see... you were imagining `PriorStateRaw` as being the JSON serialization of an entire `states.ResourceInstanceObject` here?

I guess that could work in principle, but there's currently no entry point in the `statefile` package for serializing/unserializing a single `ResourceInstanceObject`, and that'd cause the providers to depend on an aspect of the state file format, which I'd rather avoid in the interests of keeping things orthogonal.

Having a message type representing "raw state in need of upgrading" seems like a good compromise to me. I'm not sure `PriorState` speaks to me as a name for it, since we also use that term in a different way for the plan and apply methods, but maybe we could call it `RawState` to reflect that it hasn't yet been processed.

Reply:
`RawState` SGMT 👍